File: LeetCode/0018. 4Sum/ksum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
	nums.sort((a,b) => a - b);
    return kSum(nums, target, 4);
};

function kSum(nums, target, k) {
	let groups = []

	if (k === 2) return twoSum(nums, target);
	if (k > 2) 
		for (let i = 0; i < nums.length - (k-1);) {
			kSum(nums.slice(i+1, nums.length), target - nums[i], k-1)
				.forEach(group => groups.push([nums[i]].concat(group)));
			i++;
			while(i < nums.length - (k-1) && nums[i] === nums[i-1]) i++;
		}
	return groups;
}

function twoSum(nums, target) {
	let groups = [];

	let i = 0;
	let j = nums.length - 1;

	while(i < j) {
		let distance = target - nums[i] - nums[j];
		if(!distance) groups.push([nums[i], nums[j]]);
		if (distance <= 0) {
			j--;
			while(i < j && nums[j] === nums[j+1]) j--;
		}
		if (distance >= 0) {
			i++;
			while(i < j && nums[i] === nums[i-1]) i++;
		}

	}
	return groups;
}

File: LeetCode/0018. 4Sum/ksum.py
# from sys import argv
# from typing import List
class Solution:
	def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
		nums.sort()
		return self.kSum(nums, target, 4)

	def kSum(self, nums: List[int], target: int, k: int) -> List[List[int]]:
		groups = set()
		if k > 2:
			for i in range(len(nums) - (k - 1)):
				for group in self.kSum(nums[i + 1:], target - nums[i], k-1):
					group.insert(0,nums[i])
					groups.add(tuple(group))
		else:
			return self.twoSum(nums, target)

		return [list(g) for g in groups]

	def twoSum(self, nums: List[int], target: int) -> List[List[int]]:
		i = 0
		j = len(nums) - 1

		groups = set()
		while i < j:
			distance = target - nums[i] - nums[j]
			if distance is 0:
				groups.add((nums[i], nums[j]))
				i += 1
				j -= 1
				continue
			if distance > 0:
				i += 1
			else:
				j -= 1

		return [list(g) for g in groups]


# print(Solution().fourSum([int(s) for s in argv[1].split(',')], int(argv[2])))